File: state-pattern/traffic-lights/traffic-lights.ts
interface TrafficLightState {
  switchState(trafficLight: TrafficLight): void;
}

// <======= Concrete states =============>
class RedLight implements TrafficLightState {
  switchState(trafficLight: TrafficLight): void {
    console.log(`Switching from Red to Green light`);
    trafficLight.setState(new GreenLight());
  }
}

class GreenLight implements TrafficLightState {
  switchState(trafficLight: TrafficLight): void {
    console.log(`Switching from Green to Yellow light`);
    trafficLight.setState(new YellowLight());
  }
}

class YellowLight implements TrafficLightState {
  switchState(trafficLight: TrafficLight): void {
    console.log(`Switching from Yellow to Red light`);
    trafficLight.setState(new RedLight());
  }
}

// Context
class TrafficLight {
  private state: TrafficLightState;

  constructor() {
    this.state = new RedLight(); // Initial state
  }

  setState(state: TrafficLightState) {
    this.state = state;
  }

  switchState() {
    this.state.switchState(this);
  }
}

// Usage
const trafficLight = new TrafficLight();

for (let i = 0; i < 5; i++) {
  trafficLight.switchState();
}

File: state-pattern/traffic-lights/traffic-lights.rs
trait TrafficLightState {
    fn switch(self: Box<Self>, traffic_light: &mut TrafficLight);
}

// Concrete states
struct RedLight;
struct GreenLight;
struct YellowLight;

impl TrafficLightState for RedLight {
    fn switch(self: Box<Self>, traffic_light: &mut TrafficLight) {
        println!("Switching from Red to Green");
        // traffic_light.state = Box::new(GreenLight);
        traffic_light.set_state(Box::new(GreenLight));
    }
}

impl TrafficLightState for GreenLight {
    fn switch(self: Box<Self>, traffic_light: &mut TrafficLight) {
        println!("Switching from Green to Yellow");
        // traffic_light.state = Box::new(YellowLight);
        traffic_light.set_state(Box::new(YellowLight));
    }
}

impl TrafficLightState for YellowLight {
    fn switch(self: Box<Self>, traffic_light: &mut TrafficLight) {
        println!("Switching from Yellow to Red");
        // traffic_light.state = Box::new(RedLight);
        traffic_light.set_state(Box::new(RedLight));
    }
}

// Context
struct TrafficLight {
    state: Box<dyn TrafficLightState>,
}

impl TrafficLight {
    fn new() -> Self {
        Self {
            state: Box::new(RedLight), // Initially red light is on
        }
    }

    fn set_state(&mut self, state: Box<dyn TrafficLightState>) {
        self.state = state;
    }

    fn switch(&mut self) {
        let current_state = std::mem::replace(&mut self.state, Box::new(RedLight)); // Returns the old state and puts a placeholder in place, to satisfy the borrow checker
        current_state.switch(self);
    }
}

fn main() {
    let mut traffic_light = TrafficLight::new();

    for _ in 0..5 {
        // Let's simulate 6 state transitions
        traffic_light.switch();
    }
}
